File: javascript/src/local/knowledge.ts
import type { FunctionTool } from "../types/tools.js";

export type LocalKnowledgeSourceType = "transcript" | "workdir_file" | "note" | "tool_output";

export interface LocalKnowledgeScope {
  conversationId?: string;
  workspaceId?: string;
  profile?: string;
  workdir?: string;
  tags?: string[];
}

export interface LocalKnowledgeRetentionPolicy {
  /** Maximum age for transcript-derived sources, in seconds. */
  transcriptTtlSeconds?: number;
  /** Maximum age for workdir file sources that have not been refreshed, in seconds. */
  workdirTtlSeconds?: number;
  /** Maximum total stored local knowledge bytes before oldest sources are pruned. */
  maxBytes?: number;
  /** Maximum transcript-derived sources to retain. */
  maxTranscriptSources?: number;
  /** Maximum workdir file sources to retain. */
  maxWorkdirSources?: number;
  /** Remove soft-deleted sources older than this many seconds. */
  deletedTtlSeconds?: number;
}

export interface LocalKnowledgeRetrievalPolicy {
  defaultLimit?: number;
  maxLimit?: number;
  defaultContextBytes?: number;
  maxContextBytes?: number;
  /** prefer ranks active scope first; filter excludes records outside the active scope. */
  scopeMode?: "prefer" | "filter";
  /** In filter mode, allow same workspace/profile knowledge from other conversations. */
  includeConversationSiblings?: boolean;
}

export interface LocalKnowledgeIngestionPolicy {
  maxTranscriptBytes?: number;
  maxWorkdirFiles?: number;
  maxWorkdirFileBytes?: number;
  maxChunkBytes?: number;
  includeWorkdir?: boolean;
  includeTranscripts?: boolean;
}

export interface LocalKnowledgePolicy {
  enabled?: boolean;
  retention?: LocalKnowledgeRetentionPolicy;
  retrieval?: LocalKnowledgeRetrievalPolicy;
  ingestion?: LocalKnowledgeIngestionPolicy;
}

export interface LocalKnowledgeIngestMessage {
  conversationId: string;
  messageId: string;
  role: "user" | "assistant" | "system";
  kind?: "tool";
  text: string;
  scope?: LocalKnowledgeScope;
}

export interface LocalKnowledgeIngestWorkdirOptions {
  root: string;
  maxFiles?: number;
  maxBytesPerFile?: number;
  scope?: LocalKnowledgeScope;
}

export interface LocalKnowledgeSearchParams {
  query: string;
  limit?: number;
  scope?: LocalKnowledgeScope;
  /** @deprecated Use scope.conversationId. */
  conversationId?: string;
  /** @deprecated Use scope.workdir. */
  workdir?: string;
}

export interface LocalKnowledgeHit {
  id: string;
  sourceType: LocalKnowledgeSourceType;
  sourceUri: string;
  title?: string;
  text: string;
  score?: number;
  updatedAt?: number;
  metadata?: Record<string, unknown>;
}

export interface LocalKnowledgeSearchResult {
  object: "local_knowledge_search_result";
  data: LocalKnowledgeHit[];
}

export interface LocalKnowledgeContextParams extends LocalKnowledgeSearchParams {
  maxBytes?: number;
}

export interface LocalKnowledgeContext {
  hits: LocalKnowledgeHit[];
  text: string;
}

export interface LocalKnowledgeStats {
  object: "local_knowledge_stats";
  sources: number;
  chunks: number;
  bytes: number;
  deletedSources: number;
  oldestIndexedAt?: number;
  newestIndexedAt?: number;
  bySourceType?: Partial<Record<LocalKnowledgeSourceType, {
    sources: number;
    chunks: number;
    bytes: number;
  }>>;
}

export interface LocalKnowledgePruneParams {
  policy?: LocalKnowledgePolicy;
  scope?: LocalKnowledgeScope;
  dryRun?: boolean;
}

export interface LocalKnowledgePruneResult {
  object: "local_knowledge_prune_result";
  dryRun?: boolean;
  deletedSources: number;
  deletedChunks: number;
  reclaimedBytes: number;
}

export interface LocalKnowledgeForgetParams {
  conversationId?: string;
  workspaceId?: string;
  profile?: string;
  workdir?: string;
  sourceUri?: string;
  sourceType?: LocalKnowledgeSourceType;
}

export interface LocalKnowledgeService {
  ingestMessage?(message: LocalKnowledgeIngestMessage): Promise<void> | void;
  ingestWorkdir?(options: LocalKnowledgeIngestWorkdirOptions): Promise<void> | void;
  forgetConversation?(conversationId: string): Promise<void> | void;
  forget?(params: LocalKnowledgeForgetParams): Promise<void> | void;
  prune?(params?: LocalKnowledgePruneParams): Promise<LocalKnowledgePruneResult> | LocalKnowledgePruneResult;
  stats?(scope?: LocalKnowledgeScope): Promise<LocalKnowledgeStats> | LocalKnowledgeStats;
  search(params: LocalKnowledgeSearchParams): Promise<LocalKnowledgeSearchResult>;
  contextForPrompt(params: LocalKnowledgeContextParams): Promise<LocalKnowledgeContext | null>;
  dispose?(): void;
}

export interface LocalKnowledgeToolRegistry {
  readonly toolName: string;
  definitions(): FunctionTool[];
  execute(name: string, args: Record<string, unknown>): Promise<Record<string, unknown>>;
}

export function createLocalKnowledgeToolRegistry(
  service: LocalKnowledgeService,
  toolName = "local_knowledge",
): LocalKnowledgeToolRegistry {
  const definition = localKnowledgeToolDefinition(toolName);
  return {
    toolName,
    definitions: () => [{ ...definition }],
    execute: async (name, args) => {
      if (name !== toolName) throw new Error(`unknown local knowledge tool: ${name}`);
      const action = typeof args.action === "string" ? args.action : "search";
      if (action !== "search") {
        return localKnowledgeErrorResult(action, `unsupported local_knowledge action: ${action}`);
      }
      const query = typeof args.query === "string" ? args.query.trim() : "";
      if (!query) {
        return localKnowledgeErrorResult(action, "query is required");
      }
      const limit = typeof args.limit === "number" ? args.limit : undefined;
      return {
        object: "local_knowledge_result",
        action,
        result: await service.search({ query, limit }),
      };
    },
  };
}

export function localKnowledgeToolDefinition(name = "local_knowledge"): FunctionTool {
  return {
    type: "function",
    name,
    description: [
      "Search the local knowledge index for durable local context from prior transcript messages and indexed workdir files.",
      "Use this when the current task may depend on local project history, prior decisions, repo conventions, or nearby documentation.",
      "The index is local and may be incomplete; use local_workdir to verify exact current file contents before editing.",
    ].join(" "),
    parameters: {
      type: "object",
      properties: {
        action: {
          type: "string",
          enum: ["search"],
          description: "Local knowledge operation.",
        },
        query: {
          type: "string",
          description: "Natural language or keyword query.",
        },
        limit: {
          type: "integer",
          description: "Maximum number of hits.",
        },
      },
      required: ["action", "query"],
      additionalProperties: false,
    },
    strict: false,
  };
}

export function formatLocalKnowledgeContext(context: LocalKnowledgeContext): string {
  if (context.hits.length === 0) return "";
  return [
    "Local knowledge follows. It is retrieved from local transcripts and indexed local files; treat it as contextual hints and verify exact file contents with local_workdir when precision matters.",
    context.text,
  ].filter(Boolean).join("\n");
}

function localKnowledgeErrorResult(action: string, message: string): Record<string, unknown> {
  return {
    object: "local_knowledge_result",
    action,
    error: { message },
  };
}

File: python/src/agent_api/local/knowledge.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Literal, Protocol, TypedDict, cast

from agent_api.types.tools import FunctionTool, Tool

LocalKnowledgeSourceType = Literal["transcript", "workdir_file", "note", "tool_output"]


class LocalKnowledgeScope(TypedDict, total=False):
    conversationId: str
    workspaceId: str
    profile: str
    workdir: str
    tags: list[str]


class LocalKnowledgeRetentionPolicy(TypedDict, total=False):
    transcriptTtlSeconds: int
    workdirTtlSeconds: int
    maxBytes: int
    maxTranscriptSources: int
    maxWorkdirSources: int
    deletedTtlSeconds: int


class LocalKnowledgeRetrievalPolicy(TypedDict, total=False):
    defaultLimit: int
    maxLimit: int
    defaultContextBytes: int
    maxContextBytes: int
    scopeMode: Literal["prefer", "filter"]
    includeConversationSiblings: bool


class LocalKnowledgeIngestionPolicy(TypedDict, total=False):
    maxTranscriptBytes: int
    maxWorkdirFiles: int
    maxWorkdirFileBytes: int
    maxChunkBytes: int
    includeWorkdir: bool
    includeTranscripts: bool


class LocalKnowledgePolicy(TypedDict, total=False):
    enabled: bool
    retention: LocalKnowledgeRetentionPolicy
    retrieval: LocalKnowledgeRetrievalPolicy
    ingestion: LocalKnowledgeIngestionPolicy


class LocalKnowledgeHit(TypedDict, total=False):
    id: str
    sourceType: LocalKnowledgeSourceType
    sourceUri: str
    title: str
    text: str
    score: float
    updatedAt: int
    metadata: dict[str, Any]


class LocalKnowledgeSearchResult(TypedDict):
    object: Literal["local_knowledge_search_result"]
    data: list[LocalKnowledgeHit]


class LocalKnowledgeContext(TypedDict):
    hits: list[LocalKnowledgeHit]
    text: str


class LocalKnowledgeSourceStats(TypedDict):
    sources: int
    chunks: int
    bytes: int


class LocalKnowledgeStats(TypedDict, total=False):
    object: Literal["local_knowledge_stats"]
    sources: int
    chunks: int
    bytes: int
    deletedSources: int
    oldestIndexedAt: int
    newestIndexedAt: int
    bySourceType: dict[LocalKnowledgeSourceType, LocalKnowledgeSourceStats]


class LocalKnowledgePruneParams(TypedDict, total=False):
    policy: LocalKnowledgePolicy
    scope: LocalKnowledgeScope
    dryRun: bool


class LocalKnowledgePruneResult(TypedDict, total=False):
    object: Literal["local_knowledge_prune_result"]
    dryRun: bool
    deletedSources: int
    deletedChunks: int
    reclaimedBytes: int


class LocalKnowledgeForgetParams(TypedDict, total=False):
    conversationId: str
    workspaceId: str
    profile: str
    workdir: str
    sourceUri: str
    sourceType: LocalKnowledgeSourceType


class LocalKnowledgeSearchParams(TypedDict, total=False):
    query: str
    limit: int
    scope: LocalKnowledgeScope
    conversationId: str
    workdir: str


class LocalKnowledgeContextParams(LocalKnowledgeSearchParams, total=False):
    maxBytes: int


class LocalKnowledgeIngestMessage(TypedDict, total=False):
    conversationId: str
    messageId: str
    role: Literal["user", "assistant", "system"]
    kind: Literal["tool"]
    text: str
    scope: LocalKnowledgeScope


class LocalKnowledgeIngestWorkdirOptions(TypedDict, total=False):
    root: str
    maxFiles: int
    maxBytesPerFile: int
    scope: LocalKnowledgeScope


class LocalKnowledgeService(Protocol):
    def search(self, params: LocalKnowledgeSearchParams) -> LocalKnowledgeSearchResult: ...

    def context_for_prompt(self, params: LocalKnowledgeContextParams) -> LocalKnowledgeContext | None: ...

    def ingest_message(self, message: LocalKnowledgeIngestMessage) -> None: ...

    def ingest_workdir(self, options: LocalKnowledgeIngestWorkdirOptions) -> None: ...

    def forget_conversation(self, conversation_id: str) -> None: ...

    def forget(self, params: LocalKnowledgeForgetParams) -> None: ...

    def prune(self, params: LocalKnowledgePruneParams | None = None) -> LocalKnowledgePruneResult: ...

    def stats(self, scope: LocalKnowledgeScope | None = None) -> LocalKnowledgeStats: ...

    def dispose(self) -> None: ...


LOCAL_KNOWLEDGE_TOOL_DESCRIPTION = " ".join(
    [
        "Search the local knowledge index for durable local context from prior transcript messages and indexed workdir files.",
        "Use this when the current task may depend on local project history, prior decisions, repo conventions, or nearby documentation.",
        "The index is local and may be incomplete; use local_workdir to verify exact current file contents before editing.",
    ]
)


@dataclass(frozen=True)
class LocalKnowledgeToolRegistry:
    service: LocalKnowledgeService
    tool_name: str = "local_knowledge"

    def definitions(self) -> list[Tool]:
        return [cast(Tool, dict(local_knowledge_tool_definition(self.tool_name)))]

    def execute(self, name: str, args: Mapping[str, Any]) -> dict[str, Any]:
        if name != self.tool_name:
            raise ValueError(f"unknown local knowledge tool: {name}")
        action = args.get("action") if isinstance(args.get("action"), str) else "search"
        if action != "search":
            return _local_knowledge_error_result(str(action), f"unsupported local_knowledge action: {action}")
        query = args.get("query")
        query = query.strip() if isinstance(query, str) else ""
        if not query:
            return _local_knowledge_error_result("search", "query is required")
        limit = args.get("limit") if isinstance(args.get("limit"), int) else None
        params: LocalKnowledgeSearchParams = {"query": query}
        if limit is not None:
            params["limit"] = limit
        return {
            "object": "local_knowledge_result",
            "action": "search",
            "result": self.service.search(params),
        }


def create_local_knowledge_tool_registry(
    service: LocalKnowledgeService,
    *,
    tool_name: str = "local_knowledge",
) -> LocalKnowledgeToolRegistry:
    return LocalKnowledgeToolRegistry(service=service, tool_name=tool_name)


def local_knowledge_tool_definition(name: str = "local_knowledge") -> FunctionTool:
    return {
        "type": "function",
        "name": name,
        "description": LOCAL_KNOWLEDGE_TOOL_DESCRIPTION,
        "parameters": {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "enum": ["search"],
                    "description": "Local knowledge operation.",
                },
                "query": {
                    "type": "string",
                    "description": "Natural language or keyword query.",
                },
                "limit": {
                    "type": "integer",
                    "description": "Maximum number of hits.",
                },
            },
            "required": ["action", "query"],
            "additionalProperties": False,
        },
        "strict": False,
    }


def format_local_knowledge_context(context: LocalKnowledgeContext) -> str:
    if not context["hits"]:
        return ""
    return "\n".join(
        [
            "Local knowledge follows. It is retrieved from local transcripts and indexed local files; treat it as contextual hints and verify exact file contents with local_workdir when precision matters.",
            context["text"],
        ]
    )


def _local_knowledge_error_result(action: str, message: str) -> dict[str, Any]:
    return {
        "object": "local_knowledge_result",
        "action": action,
        "error": {"message": message},
    }
